File: app/src/main/java/com/moisesborges/tvaddict/data/ShowsRepositoryImpl.java
package com.moisesborges.tvaddict.data;

import android.support.annotation.NonNull;

import com.moisesborges.tvaddict.models.CastMember;
import com.moisesborges.tvaddict.models.Episode;
import com.moisesborges.tvaddict.models.Season;
import com.moisesborges.tvaddict.models.Show;
import com.moisesborges.tvaddict.net.TvMazeApi;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Moisés on 16/04/2017.
 */

public class ShowsRepositoryImpl implements ShowsRepository {

    private final TvMazeApi mTvMazeApi;
    private final ShowDb mShowDb;
    private final EpisodesDb mEpisodesDb;
    private final SeasonDb mSeasonDb;

    public ShowsRepositoryImpl(@NonNull TvMazeApi tvMazeApi,
                               @NonNull ShowDb showDb,
                               @NonNull EpisodesDb episodesDb,
                               @NonNull SeasonDb seasonDb) {
        mTvMazeApi = tvMazeApi;
        mShowDb = showDb;
        mEpisodesDb = episodesDb;
        mSeasonDb = seasonDb;
    }

    @Override
    public Single<List<Show>> getShows() {
        return mTvMazeApi.listShows(0);
    }

    @Override
    public Single<Show> getFullShowInfo(int showId) {
        return mTvMazeApi.fetchShowFullInfo(showId)
                .doOnSuccess(addShowIdToEmbeddedData());
    }

    @NonNull
    private Consumer<Show> addShowIdToEmbeddedData() {
        return show -> {
            for (Season season : show.getSeasons()) {
                season.setShowId(show.getId());
            }
            for (Episode episode : show.getEpisodes()) {
                episode.setShowId(show.getId());
            }
            for (CastMember castMember : show.getCast()) {
                castMember.setShowId(show.getId());
            }
        };
    }

    @Override
    public Completable saveShow(@NonNull Show show) {
        return mShowDb.save(show);
    }

    @Override
    public Completable saveShowEmbeddedData(@NonNull Show show) {
        return Completable.fromRunnable(() -> {
            mSeasonDb.saveSeasons(show.getSeasons());
            mEpisodesDb.saveEpisodes(show.getEpisodes());
        });
    }

    @Override
    public Single<List<Show>> getSavedShows() {
        return mShowDb.findAllShows();
    }

    @Override
    public Single<Show> getSavedShow(int showId) {
        return mShowDb.findShow(showId);
    }

    @Override
    public Completable removeShow(int showId) {
        return mShowDb.remove(showId);
    }
}

File: app/src/main/java/com/moisesborges/tvaddict/data/ShowsRepositoryImpl.kt
package com.moisesborges.tvaddict.data

import com.moisesborges.tvaddict.models.Show
import com.moisesborges.tvaddict.models.UpcomingEpisode
import com.moisesborges.tvaddict.net.TvMazeApi
import io.reactivex.Single
import io.reactivex.functions.Function

/**
 * Created by Moisés on 16/04/2017.
 */

class ShowsRepositoryImpl(private val tvMazeApi: TvMazeApi,
                          private val showDb: ShowDb) : ShowsRepository {

    override fun loadShows(page: Int): Single<List<Show>> {
        return tvMazeApi.listShows(page)
    }

    override fun getFullShowInfo(showId: Int): Single<Show> {
        return tvMazeApi.fetchShowFullInfo(showId)
    }

    override fun saveFullShowInfo(show: Show): Single<Show> {
        return getFullShowInfo(showId = show.id)
                .onErrorReturn { Show.NOT_FOUND }
                .flatMap { fullShowInfo -> saveShow(if (fullShowInfo != Show.NOT_FOUND) fullShowInfo else show) }

    }

    override fun saveShow(show: Show): Single<Show> {
        return showDb.save(show)
                .toSingle { show }
    }

    override fun getSavedShows(): Single<List<Show>> = showDb.findAllShows()

    override fun getSavedShow(showId: Int): Single<Show> {
        return showDb.findShow(showId)
    }

    override fun removeShow(showId: Int): Single<Show> {
        return showDb.remove(showId)
                .toSingle { Show.NOT_FOUND }
    }

    override fun updateShow(show: Show): Single<Show> {
        return showDb.update(show)
    }

    override fun fetchUpcomingEpisodes(): Single<List<UpcomingEpisode>> {
        return showDb.findAllShows()
                .map { shows -> upcomingEpisodes(shows) }
    }

    private fun upcomingEpisodes(shows: List<Show>): List<UpcomingEpisode> {
        return shows.sortedWith(compareBy({ it.name }))
                .associate { show ->
                    val episodes = show.episodes.filter { episode -> !episode.wasWatched() }
                            .firstOrNull()
                    Pair(show, episodes)
                }
                .filter { pair -> pair.value != null }
                .map { pair -> UpcomingEpisode(pair.key, pair.value!!) }
                .requireNoNulls()
    }

    override fun searchShows(showName: String): Single<List<Show>> {
        return tvMazeApi.searchShows(showName)
                .map { showsInfo -> showsInfo.map { showInfo -> showInfo.show } }
    }
}
